### PR TITLE
Hadoop-BAM: add support for efficiently loading a subset of a BAM

### PR DESCRIPTION
**Pull request by tomwhite**
Addresses #1399

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 1]
```diff
@@ -92,17 +93,23 @@ public ReadsSparkSource(final JavaSparkContext ctx, final ValidationStringency v
...
             SAMRecord sam = v1._2().get();
-            if (samRecordOverlaps(sam, intervals)) {
-                return (GATKRead)SAMRecordToGATKReadAdapter.headerlessReadAdapter(sam);
+            if (isBam || samRecordOverlaps(sam, intervals)) { // don't check overlaps for BAM since it is done by input format
+                return (GATKRead) SAMRecordToGATKReadAdapter.headerlessReadAdapter(sam);
             }
             return null;
-
         }).filter(v1 -> v1 != null);
     }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: There is a TODO below for the `samRecordOverlaps()` method that says "remove this method when https://github.com/broadinstitute/hellbender/issues/559 is fixed", but that ticket has been closed as resolved. Can you create a new ticket "Use `IntervalsSkipList` in `ReadSparkSource.samRecordOverlaps()`, update the docs for the method to point to that ticket, and also add a note to the `samRecordOverlaps()` that it is only used as a fallback option for formats that don't support query-by-interval natively at the Hadoop-BAM layer?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 2]
```diff
@@ -92,17 +93,23 @@ public ReadsSparkSource(final JavaSparkContext ctx, final ValidationStringency v
 
         setHadoopBAMConfigurationProperties(readFileName, referencePath);
 
+        boolean isBam = IOUtils.isBamFileName(readFileName);
+        if (isBam && intervals != null && !intervals.isEmpty()) {
+            BAMInputFormat.setIntervals(conf, intervals);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Have we tested `BAMInputFormat.setIntervals()` with a large interval list like `ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/b37/Broad.human.exome.b37.interval_list.gz`? I wonder whether there might be performance issues, given that it converts all the intervals into one huge String and re-parses them later.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 3]
```diff
@@ -187,6 +186,21 @@ public void testCRAMReferenceFromHDFS() throws Exception {
...
+    @Test
+    public void testIntervals(){
+        JavaSparkContext ctx = SparkContextFactory.getTestSparkContext();
+        ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        List<SimpleInterval> intervals =
+                ImmutableList.of(new SimpleInterval("17", 69010, 69040), new SimpleInterval("17", 69910, 69920));
+        JavaRDD<GATKRead> reads = readSource.getParallelReads(NA12878_chr17_1k_BAM, null, intervals);
+
+        SamReaderFactory samReaderFactory = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT);
+        SamReader samReader = samReaderFactory.open(new File(NA12878_chr17_1k_BAM));
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: try-with-resources

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java [hunk 4]
```diff
@@ -187,6 +186,21 @@ public void testCRAMReferenceFromHDFS() throws Exception {
...
+        int seqIndex = samReader.getFileHeader().getSequenceIndex("17");
+        SAMRecordIterator query = samReader.query(new QueryInterval[]{new QueryInterval(seqIndex, 69010, 69040), new QueryInterval(seqIndex, 69910, 69920)}, false);
+        Assert.assertEquals(reads.count(), Iterators.size(query));
+    }
+
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: It's probably worth adding a test case with the large interval list I linked to above to make sure there are no performance problems there (should probably be checked in to `src/test/resources/large` using `git lfs`). Can use the methods in `IntervalArgumentCollection` to load the interval list.

**Comment by droazen** [comment]
:+1: merge after addressing comments. If the test I requested is too slow to be part of our test suite, should create a ticket "Fix performance issues in `BAMInputFormat.setIntervals()` with large interval lists"

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 1 above)
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: Opened https://github.com/broadinstitute/gatk/issues/1531 for this.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java (hunk 3 above)
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java]: Fixed.

**Comment by tomwhite** [comment]
I addressed the comments except the one about the large intervals file. I agree that this is a problem and we need to test for this case (which I opened #1532 for). It looks like `IntervalArgumentCollection` doesn't support reading from a file containing a list of intervals though, despite the javadoc for `getIntervalStrings()`. There's probably a bigger issue for handling large intervals for other tools, since there are no tests for that either.

**Comment by droazen** [comment]
@tomwhite `IntervalArgumentCollection` does work with interval files (tried it just now) -- you just need to make sure the extension is one of `IntervalUtils.INTERVAL_FILE_EXTENSIONS` (won't work with gzipped files, need to extract those first)

**Comment by tomwhite** [comment]
@droazen ah ok. My mistake.

**Pull request merged**